src/distributed-callback-queue.js

```diff
@@ -209,14 +222,21 @@ class DistributedCallbackQueue {
 
     let onCompleted;
     try {
-      onCompleted = await this.push(suffix, onJobCompleted, timeout);
+      const pushPromise = this.push(suffix, onJobCompleted, timeout);
+      onCompleted = await Promise.race([
+        pushPromise,
+        jobAbortPromise,
+      ]);
```

Comment:
perform a `Promise.race` between (1) acquiring the `ioredis-lock` coming back from #push and (2) a local Promise that will reject if the Job receives an error before the lock gets acquired

Reply:
What would happen to jobAbortPromise if `pushPromise` resolves? Could it result in a memory leak?

Technically what would generally happen is that once pushPromise resolves - then jobAbortPromise would never have reject/resolve called. That would, in turn, have the resolve/reject functions retained forever. 

This is just a thought, thought, and I, frankly, don't know where I'm write or wrong here as we'd technically not have any references to the jobAbortPromise anywhere and it should be GCed. But maybe its retained internally until resolve/reject is called and, therefore, would leak.

Reply:
reading through what you've written -- yeah, that's my line of thinking of well.  you're using `bluebird`.  those guys are good.  if their impl leaked through the GC like that, i imagine they would have fixed it *years* ago

and you'd imagine there are definitely Promises that never get resolved or rejected in peoples' code in the wild, so its not a strange usage pattern

i'm rationalizing here ... but i think it's safe